### ISSUE TITLE
Build fails on latest Linux due to header inclusion change in gcc/clang 13

### ISSUE THREAD

**Issue by proatgram** [google/nearby#1673]
### Project

Nearby Connections

### Language

C++

### OS Platform

Linux

### What happened?

When building using the README with the Bazel build system, things fail to build due to C integer types not being "defined" (file not being included). This happens in some of the code in this repository, in internal/crypto/sha2.h, internal/crypto/nearby_base.h, internal/crypto/random.h and in third party libraries such as abseil-cpp. This is due to [this](https://gcc.gnu.org/gcc-13/porting_to.html#header-dep-changes) change in GCC 13.

With the files in this repository, all that would need to be changed is adding  `#include <cstdint>` or `#include <stdint.h>` and for abseil-cpp, increasing the submodule commit to a commit equal or greater to commit [b957f0c](https://github.com/abseil/abseil-cpp/commit/b957f0ccd00481cd4fd663d8320aa02ae0564f18).

### What did you expect to happen?

I expected it to build successfully without error.

### How can we reproduce it (as minimally and precisely as possible)?

1. Update/use GCC 13 or higher.
2. Follow the steps in the Nearby Connections README.

### How often does this bug happen?

Every time

### Relevant log output

```shell
ERROR: /home/thetimbrick/google/nearby/internal/crypto/BUILD:23:11: Compiling internal/crypto/sha2.cc failed: (Exit 1): clang-15 failed: error executing command (from target //internal/crypto:crypto)
  (cd /home/thetimbrick/.cache/bazel/_bazel_thetimbrick/54eafa58feb1122e78a2e515ed11b705/execroot/__main__ && \
  exec env - \
    BAZEL_CXXOPTS='-std=c++17' \
    PATH=/usr/local/sbin:/usr/local/bin:/usr/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl:/usr/lib/rustup/bin \
    PWD=/proc/self/cwd \
  /usr/bin/clang-15 -U_FORTIFY_SOURCE -fstack-protector -Wall -Wthread-safety -Wself-assign -Wunused-but-set-parameter -Wno-free-nonheap-object -fcolor-diagnostics -fno-omit-frame-pointer '-std=c++17' -MD -MF bazel-out/k8-fastbuild/bin/internal/crypto/_objs/crypto/sha2.pic.d '-frandom-seed=bazel-out/k8-fastbuild/bin/internal/crypto/_objs/crypto/sha2.pic.o' -fPIC '-DBAZEL_CURRENT_REPOSITORY=""' -iquote . -iquote bazel-out/k8-fastbuild/bin -iquote external/boringssl -iquote bazel-out/k8-fastbuild/bin/external/boringssl -iquote external/com_google_absl -iquote bazel-out/k8-fastbuild/bin/external/com_google_absl -isystem external/boringssl/src/include -isystem bazel-out/k8-fastbuild/bin/external/boringssl/src/include -DCRYPTO_IMPLEMENTATION -Ithird_party -no-canonical-prefixes -Wno-builtin-macro-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"' '-D__TIME__="redacted"' -c internal/crypto/sha2.cc -o bazel-out/k8-fastbuild/bin/internal/crypto/_objs/crypto/sha2.pic.o)
# Configuration: a7b732125cae20632358513ffa9ec1499ce0ba5122adaa7480c6548305ecada8
# Execution platform: @local_config_platform//:host
In file included from internal/crypto/sha2.cc:15:
./internal/crypto/sha2.h:36:26: error: use of undeclared identifier 'uint8_t'
CRYPTO_EXPORT std::array<uint8_t, kSHA256Length> SHA256Hash(
                         ^
./internal/crypto/sha2.h:37:22: error: unknown type name 'uint8_t'
    absl::Span<const uint8_t> input);
                     ^
2 errors generated.
Target //connections:core failed to build
INFO: Elapsed time: 413.084s, Critical Path: 44.32s
INFO: 375 processes: 107 internal, 268 local.
FAILED: Build did NOT complete successfully


ERROR: /home/thetimbrick/google/nearby/internal/crypto/BUILD:23:11: Compiling internal/crypto/nearby_base.cc failed: (Exit 1): clang-15 failed: error executing command (from target //internal/crypto:crypto)
  (cd /home/thetimbrick/.cache/bazel/_bazel_thetimbrick/54eafa58feb1122e78a2e515ed11b705/execroot/__main__ && \
  exec env - \
    BAZEL_CXXOPTS='-std=c++17' \
    PATH=/usr/local/sbin:/usr/local/bin:/usr/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl:/usr/lib/rustup/bin \
    PWD=/proc/self/cwd \
  /usr/bin/clang-15 -U_FORTIFY_SOURCE -fstack-protector -Wall -Wthread-safety -Wself-assign -Wunused-but-set-parameter -Wno-free-nonheap-object -fcolor-diagnostics -fno-omit-frame-pointer '-std=c++17' -MD -MF bazel-out/k8-fastbuild/bin/internal/crypto/_objs/crypto/nearby_base.pic.d '-frandom-seed=bazel-out/k8-fastbuild/bin/internal/crypto/_objs/crypto/nearby_base.pic.o' -fPIC '-DBAZEL_CURRENT_REPOSITORY=""' -iquote . -iquote bazel-out/k8-fastbuild/bin -iquote external/boringssl -iquote bazel-out/k8-fastbuild/bin/external/boringssl -iquote external/com_google_absl -iquote bazel-out/k8-fastbuild/bin/external/com_google_absl -isystem external/boringssl/src/include -isystem bazel-out/k8-fastbuild/bin/external/boringssl/src/include -DCRYPTO_IMPLEMENTATION -Ithird_party -no-canonical-prefixes -Wno-builtin-macro-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"' '-D__TIME__="redacted"' -c internal/crypto/nearby_base.cc -o bazel-out/k8-fastbuild/bin/internal/crypto/_objs/crypto/nearby_base.pic.o)
# Configuration: a7b732125cae20632358513ffa9ec1499ce0ba5122adaa7480c6548305ecada8
# Execution platform: @local_config_platform//:host
In file included from internal/crypto/nearby_base.cc:15:
./internal/crypto/nearby_base.h:33:18: error: unknown type name 'uint8_t'
absl::Span<const uint8_t> as_bytes(absl::Span<T> s) noexcept {
                 ^
./internal/crypto/nearby_base.h:34:34: error: unknown type name 'uint8_t'
  return {reinterpret_cast<const uint8_t*>(s.data()), sizeof(T) * s.size()};
                                 ^
2 errors generated.
Target //connections:core failed to build
INFO: Elapsed time: 351.377s, Critical Path: 350.73s
INFO: 548 processes: 63 internal, 485 local.
FAILED: Build did NOT complete successfully


ERROR: /home/thetimbrick/google/nearby/internal/crypto/BUILD:23:11: Compiling internal/crypto/random.cc failed: (Exit 1): clang-15 failed: error executing command (from target //internal/crypto:crypto)
  (cd /home/thetimbrick/.cache/bazel/_bazel_thetimbrick/54eafa58feb1122e78a2e515ed11b705/execroot/__main__ && \
  exec env - \
    BAZEL_CXXOPTS='-std=c++17' \
    PATH=/usr/local/sbin:/usr/local/bin:/usr/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl:/usr/lib/rustup/bin \
    PWD=/proc/self/cwd \
  /usr/bin/clang-15 -U_FORTIFY_SOURCE -fstack-protector -Wall -Wthread-safety -Wself-assign -Wunused-but-set-parameter -Wno-free-nonheap-object -fcolor-diagnostics -fno-omit-frame-pointer '-std=c++17' -MD -MF bazel-out/k8-fastbuild/bin/internal/crypto/_objs/crypto/random.pic.d '-frandom-seed=bazel-out/k8-fastbuild/bin/internal/crypto/_objs/crypto/random.pic.o' -fPIC '-DBAZEL_CURRENT_REPOSITORY=""' -iquote . -iquote bazel-out/k8-fastbuild/bin -iquote external/boringssl -iquote bazel-out/k8-fastbuild/bin/external/boringssl -iquote external/com_google_absl -iquote bazel-out/k8-fastbuild/bin/external/com_google_absl -isystem external/boringssl/src/include -isystem bazel-out/k8-fastbuild/bin/external/boringssl/src/include -DCRYPTO_IMPLEMENTATION -Ithird_party -no-canonical-prefixes -Wno-builtin-macro-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"' '-D__TIME__="redacted"' -c internal/crypto/random.cc -o bazel-out/k8-fastbuild/bin/internal/crypto/_objs/crypto/random.pic.o)
# Configuration: a7b732125cae20632358513ffa9ec1499ce0ba5122adaa7480c6548305ecada8
# Execution platform: @local_config_platform//:host
In file included from internal/crypto/random.cc:15:
./internal/crypto/random.h:37:41: error: use of undeclared identifier 'uint8_t'
CRYPTO_EXPORT void RandBytes(absl::Span<uint8_t> bytes);
                                        ^
1 error generated.
Target //connections:core failed to build
INFO: Elapsed time: 4.746s, Critical Path: 4.36s
INFO: 15 processes: 6 internal, 9 local.
FAILED: Build did NOT complete successfully


In file included from /home/thetimbrick/.cache/bazel/_bazel_thetimbrick/54eafa58feb1122e78a2e515ed11b705/execroot/__main__/external/com_google_ukey2/third_party/absl/absl/strings/internal/str_format/extension.cc:16:
/home/thetimbrick/.cache/bazel/_bazel_thetimbrick/54eafa58feb1122e78a2e515ed11b705/execroot/__main__/external/com_google_ukey2/third_party/absl/absl/strings/internal/str_format/extension.h:34:35: error: unknown type name 'uint8_t'
enum class FormatConversionChar : uint8_t;
                                  ^
/home/thetimbrick/.cache/bazel/_bazel_thetimbrick/54eafa58feb1122e78a2e515ed11b705/execroot/__main__/external/com_google_ukey2/third_party/absl/absl/strings/internal/str_format/extension.h:35:38: error: unknown type name 'uint64_t'
enum class FormatConversionCharSet : uint64_t;
                                     ^
/home/thetimbrick/.cache/bazel/_bazel_thetimbrick/54eafa58feb1122e78a2e515ed11b705/execroot/__main__/external/com_google_ukey2/third_party/absl/absl/strings/internal/str_format/extension.h:173:21: error: unknown type name 'uint8_t'
  enum class Enum : uint8_t {
                    ^
/home/thetimbrick/.cache/bazel/_bazel_thetimbrick/54eafa58feb1122e78a2e515ed11b705/execroot/__main__/external/com_google_ukey2/third_party/absl/absl/strings/internal/str_format/extension.h:327:19: error: unknown type name 'uint64_t'
      static_cast<uint64_t>(a) |
                  ^
/home/thetimbrick/.cache/bazel/_bazel_thetimbrick/54eafa58feb1122e78a2e515ed11b705/execroot/__main__/external/com_google_ukey2/third_party/absl/absl/strings/internal/str_format/extension.h:328:19: error: unknown type name 'uint64_t'
      static_cast<uint64_t>(FormatConversionCharSetUnion(rest...)));
                  ^
/home/thetimbrick/.cache/bazel/_bazel_thetimbrick/54eafa58feb1122e78a2e515ed11b705/execroot/__main__/external/com_google_ukey2/third_party/absl/absl/strings/internal/str_format/extension.h:331:11: error: unknown type name 'uint64_t'
constexpr uint64_t FormatConversionCharToConvInt(FormatConversionChar c) {
          ^
/home/thetimbrick/.cache/bazel/_bazel_thetimbrick/54eafa58feb1122e78a2e515ed11b705/execroot/__main__/external/com_google_ukey2/third_party/absl/absl/strings/internal/str_format/extension.h:332:10: error: use of undeclared identifier 'uint64_t'
  return uint64_t{1} << (1 + static_cast<uint8_t>(c));
         ^
/home/thetimbrick/.cache/bazel/_bazel_thetimbrick/54eafa58feb1122e78a2e515ed11b705/execroot/__main__/external/com_google_ukey2/third_party/absl/absl/strings/internal/str_format/extension.h:332:18: error: expected ';' after return statement
  return uint64_t{1} << (1 + static_cast<uint8_t>(c));
                 ^
                 ;
/home/thetimbrick/.cache/bazel/_bazel_thetimbrick/54eafa58feb1122e78a2e515ed11b705/execroot/__main__/external/com_google_ukey2/third_party/absl/absl/strings/internal/str_format/extension.h:335:11: error: unknown type name 'uint64_t'
constexpr uint64_t FormatConversionCharToConvInt(char conv) {
          ^
/home/thetimbrick/.cache/bazel/_bazel_thetimbrick/54eafa58feb1122e78a2e515ed11b705/execroot/__main__/external/com_google_ukey2/third_party/absl/absl/strings/internal/str_format/extension.h:348:35: error: no return statement in constexpr function
constexpr FormatConversionCharSet FormatConversionCharToConvValue(char conv) {
                                  ^
/home/thetimbrick/.cache/bazel/_bazel_thetimbrick/54eafa58feb1122e78a2e515ed11b705/execroot/__main__/external/com_google_ukey2/third_party/absl/absl/strings/internal/str_format/extension.h:357:3: error: default initialization of an object of const type 'const absl::FormatConversionCharSet'
  ABSL_INTERNAL_CONVERSION_CHARS_EXPAND_(ABSL_INTERNAL_CHAR_SET_CASE, )
  ^
/home/thetimbrick/.cache/bazel/_bazel_thetimbrick/54eafa58feb1122e78a2e515ed11b705/execroot/__main__/external/com_google_ukey2/third_party/absl/absl/strings/internal/str_format/extension.h:147:9: note: expanded from macro 'ABSL_INTERNAL_CONVERSION_CHARS_EXPAND_'
  X_VAL(c) X_SEP X_VAL(s) X_SEP \
        ^
/home/thetimbrick/.cache/bazel/_bazel_thetimbrick/54eafa58feb1122e78a2e515ed11b705/execroot/__main__/external/com_google_ukey2/third_party/absl/absl/strings/internal/str_format/extension.h:357:3: error: default initialization of an object of const type 'const absl::FormatConversionCharSet'
/home/thetimbrick/.cache/bazel/_bazel_thetimbrick/54eafa58feb1122e78a2e515ed11b705/execroot/__main__/external/com_google_ukey2/third_party/absl/absl/strings/internal/str_format/extension.h:147:24: note: expanded from macro 'ABSL_INTERNAL_CONVERSION_CHARS_EXPAND_'
  X_VAL(c) X_SEP X_VAL(s) X_SEP \
                       ^
/home/thetimbrick/.cache/bazel/_bazel_thetimbrick/54eafa58feb1122e78a2e515ed11b705/execroot/__main__/external/com_google_ukey2/third_party/absl/absl/strings/internal/str_format/extension.h:357:3: error: default initialization of an object of const type 'const absl::FormatConversionCharSet'
/home/thetimbrick/.cache/bazel/_bazel_thetimbrick/54eafa58feb1122e78a2e515ed11b705/execroot/__main__/external/com_google_ukey2/third_party/absl/absl/strings/internal/str_format/extension.h:149:9: note: expanded from macro 'ABSL_INTERNAL_CONVERSION_CHARS_EXPAND_'
  X_VAL(d) X_SEP X_VAL(i) X_SEP X_VAL(o) X_SEP \
        ^
/home/thetimbrick/.cache/bazel/_bazel_thetimbrick/54eafa58feb1122e78a2e515ed11b705/execroot/__main__/external/com_google_ukey2/third_party/absl/absl/strings/internal/str_format/extension.h:357:3: error: default initialization of an object of const type 'const absl::FormatConversionCharSet'
/home/thetimbrick/.cache/bazel/_bazel_thetimbrick/54eafa58feb1122e78a2e515ed11b705/execroot/__main__/external/com_google_ukey2/third_party/absl/absl/strings/internal/str_format/extension.h:149:24: note: expanded from macro 'ABSL_INTERNAL_CONVERSION_CHARS_EXPAND_'
  X_VAL(d) X_SEP X_VAL(i) X_SEP X_VAL(o) X_SEP \
                       ^
/home/thetimbrick/.cache/bazel/_bazel_thetimbrick/54eafa58feb1122e78a2e515ed11b705/execroot/__main__/external/com_google_ukey2/third_party/absl/absl/strings/internal/str_format/extension.h:357:3: error: default initialization of an object of const type 'const absl::FormatConversionCharSet'
/home/thetimbrick/.cache/bazel/_bazel_thetimbrick/54eafa58feb1122e78a2e515ed11b705/execroot/__main__/external/com_google_ukey2/third_party/absl/absl/strings/internal/str_format/extension.h:149:39: note: expanded from macro 'ABSL_INTERNAL_CONVERSION_CHARS_EXPAND_'
  X_VAL(d) X_SEP X_VAL(i) X_SEP X_VAL(o) X_SEP \
                                      ^
/home/thetimbrick/.cache/bazel/_bazel_thetimbrick/54eafa58feb1122e78a2e515ed11b705/execroot/__main__/external/com_google_ukey2/third_party/absl/absl/strings/internal/str_format/extension.h:357:3: error: default initialization of an object of const type 'const absl::FormatConversionCharSet'
/home/thetimbrick/.cache/bazel/_bazel_thetimbrick/54eafa58feb1122e78a2e515ed11b705/execroot/__main__/external/com_google_ukey2/third_party/absl/absl/strings/internal/str_format/extension.h:150:9: note: expanded from macro 'ABSL_INTERNAL_CONVERSION_CHARS_EXPAND_'
  X_VAL(u) X_SEP X_VAL(x) X_SEP X_VAL(X) X_SEP \
        ^
/home/thetimbrick/.cache/bazel/_bazel_thetimbrick/54eafa58feb1122e78a2e515ed11b705/execroot/__main__/external/com_google_ukey2/third_party/absl/absl/strings/internal/str_format/extension.h:357:3: error: default initialization of an object of const type 'const absl::FormatConversionCharSet'
/home/thetimbrick/.cache/bazel/_bazel_thetimbrick/54eafa58feb1122e78a2e515ed11b705/execroot/__main__/external/com_google_ukey2/third_party/absl/absl/strings/internal/str_format/extension.h:150:24: note: expanded from macro 'ABSL_INTERNAL_CONVERSION_CHARS_EXPAND_'
  X_VAL(u) X_SEP X_VAL(x) X_SEP X_VAL(X) X_SEP \
                       ^
/home/thetimbrick/.cache/bazel/_bazel_thetimbrick/54eafa58feb1122e78a2e515ed11b705/execroot/__main__/external/com_google_ukey2/third_party/absl/absl/strings/internal/str_format/extension.h:357:3: error: default initialization of an object of const type 'const absl::FormatConversionCharSet'
/home/thetimbrick/.cache/bazel/_bazel_thetimbrick/54eafa58feb1122e78a2e515ed11b705/execroot/__main__/external/com_google_ukey2/third_party/absl/absl/strings/internal/str_format/extension.h:150:39: note: expanded from macro 'ABSL_INTERNAL_CONVERSION_CHARS_EXPAND_'
  X_VAL(u) X_SEP X_VAL(x) X_SEP X_VAL(X) X_SEP \
                                      ^
/home/thetimbrick/.cache/bazel/_bazel_thetimbrick/54eafa58feb1122e78a2e515ed11b705/execroot/__main__/external/com_google_ukey2/third_party/absl/absl/strings/internal/str_format/extension.h:357:3: error: default initialization of an object of const type 'const absl::FormatConversionCharSet'
/home/thetimbrick/.cache/bazel/_bazel_thetimbrick/54eafa58feb1122e78a2e515ed11b705/execroot/__main__/external/com_google_ukey2/third_party/absl/absl/strings/internal/str_format/extension.h:152:9: note: expanded from macro 'ABSL_INTERNAL_CONVERSION_CHARS_EXPAND_'
  X_VAL(f) X_SEP X_VAL(F) X_SEP X_VAL(e) X_SEP X_VAL(E) X_SEP \
        ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
5 warnings and 20 errors generated.
make[2]: *** [third_party/absl/absl/strings/CMakeFiles/absl_str_format_internal.dir/build.make:104: third_party/absl/absl/strings/CMakeFiles/absl_str_format_internal.dir/internal/str_format/extension.cc.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:2408: third_party/absl/absl/strings/CMakeFiles/absl_str_format_internal.dir/all] Error 2
make: *** [Makefile:146: all] Error 2
_____ END BUILD LOGS _____
rules_foreign_cc: Build wrapper script location: bazel-out/k8-fastbuild/bin/connections/implementation/ukey2_foreign_cc/wrapper_build_script.sh
rules_foreign_cc: Build script location: bazel-out/k8-fastbuild/bin/connections/implementation/ukey2_foreign_cc/build_script.sh
rules_foreign_cc: Build log location: bazel-out/k8-fastbuild/bin/connections/implementation/ukey2_foreign_cc/CMake.log

Target //connections:core failed to build
INFO: Elapsed time: 494.212s, Critical Path: 481.14s
INFO: 4 processes: 2 internal, 2 local.
FAILED: Build did NOT complete successfully
```

### Anything else we need to know?

Adding `#include <cstdint>` or `#include <stdint.h>` to internal/crypto/sha2.h, internal/crypto/nearby_base.h, and internal/crypto/random.h should fix the issue in the internal sources.

Updating the commit for the abseil-cpp submodule to a version greater than or equal to [b957f0c](https://github.com/abseil/abseil-cpp/commit/b957f0ccd00481cd4fd663d8320aa02ae0564f18) in third_party/absl should solve that same issue for this dependency.

In the third party library ukey2, the same thing is happening with the C integer inclusion in their submodule for abseil-cpp, although that is a problem with that project and not this one. I am trying to build ukey2 by itself but it doesn't seem to want to build when I am not building using the build for Nearby Connections. The solution to this would to also increase the commit hash for it's abseil-cpp dependency to the one that fixes this issue. There is a PR in google/ukey2 that would remove the need for abseil-cpp that would solve that issue for that repository, but the PR and issue have been stagnant.

This bug might be relevant in other projects in this repository but they did not have instructions on how to build them, and since I'm not familiar with the Bazel build system I have not been able to build and test them yet.

The current GCC version in Ubuntu Latest is GCC 10 it looks like, so the issue will most likely not be reproducable in that enviornment.

**Comment by proatgram** [comment]
I have also created some PR's and issues that fix the compilation of Ukey2 on the latest GCC/Clang. Ukey2's dependency securemessage was failing to build due to a bug in it's gtest submodule. [This PR](https://github.com/google/securemessage/pull/8) fixes that. If that gets merged then Ukey2 will need to update it's securemessage submodule in order to reflect that ([issue](https://github.com/google/ukey2/issues/18)) and their abseil-cpp submodule to reflect the GCC/Clang 13 changes ([issue](https://github.com/google/ukey2/issues/19)). And if that gets merged, a submodule update in this repository would be necessary to bring the bug fixes from Ukey2 to here.

**Comment by bourdakos1** [comment]
Thanks @proatgram 😁 @suetfei do you mind taking a look at these PRs?

**Comment by proatgram** [comment]
@bourdakos1 No problem! 😁 Just to note, the PR's I included are to dependency repositories. I can create one that will fix the C integer inclusion for the crypto headers in this repository and the abseil-cpp fix but I will need to wait until Ukey2 gets updated before I can create a PR in this repository to update the submodule for Ukey2.

**Comment by proatgram** [comment]
I also missed that securemessage is also a dependency of Nearby as well, and not just Ukey2, so the submodule for that will need to be updated if [this PR](https://github.com/google/securemessage/pull/8) is merged into securemessage.

**Comment by suetfei** [comment]
Thanks @proatgram for the bug report. Currently Github Action runners are still using clang 14.0. We may wait for that to upgrade before we handle this issue.

In the meanwhile, we are considering some refactoring so all nearby dependencies have the same "source". Currently this is not the case because the Bazel build invokes cmake and they can point to different versions of the same DEP, i.e. absl. We are evaluating different proposals.

**Comment by proatgram** [comment]
Thanks @suetfei for the insight. In the meantime while GitHub action runners are using Clang 14 and you and your team are evaluating proposals, how should I go about building and developing this on my own system which has the latest Clang/GCC? I was planning on tinkering with creating some mediums for Linux but first I need a working compilation.

**Comment by suetfei** [comment]
Could you install an older version of clang and use that to build Nearby? Currently you are using "  /usr/bin/clang-15". Sorry about the inconvenience.

**Comment by proatgram** [comment]
I will experiment around with it, although it might conflict with already installed packages that depend on the latest Clang. Thank you for helping!

Another thing I thought to ask as you mentioned doing refactoring was if there were plans to implement a way to use already installed libraries and dependencies if they are already installed on your system instead of building them from source. There are a couple of reasons behind doing this, one of which is to reduce bundled packages and libraries which package maintainers don't really like and to have just one install and one version of the dependency installed on the system.

Thanks for the help and insight you are providing! 😁

**Comment by suetfei** [comment]
Good news! The team has been working on it and we should have a solution in a couple days. The solution will speed up the Build time by 7 minutes because  the time consuming foreign_cc part (lots of copying!!) will be gone. 

I can see the appeal of using "already installed libraries and dependencies" from a user's prospective. However, from a distribution perspective, we want the Build process to be deterministic. That is, every user runs exactly the same code. And the only way to ensure this is to build from source. Otherwise we will have situation that one user can build but the other can not and that is hard to debug. 

Thanks very much for the suggestion! Please keep your suggestions/PRs coming. We really appreciate it.

**Comment by proatgram** [comment]
That's good! Thank you and your team for all of the hard work that is being put in to make this work and be a better experience!

Ahh I see, that is a valid point. My thought process was that for example if there were breaking issues in a dependency in this project, for example abseil-cpp, that that would have already been addressed by the maintainers and there wouldn't be an issue with it. Here's an [article](https://fedoraproject.org/wiki/Bundled_Libraries) explaining some more of what I was thinking when I suggested this. And I totally get your reasoning behind it as you said, I have had many times where that has happened.

No problem! I very much enjoy contributing and helping in any way possible! I have a [commit](https://github.com/proatgram/nearby/commit/6e6929ac272330946b1f1f9ca1c18c88f1c67fc6) on a fork that fixes the issues that I was having with newer versions of Clang and GCC, which just adds an include to a couple files and increments the version of abseil to a commit that fixes the same issue in that project; although you mentioned your team wanted to wait to handle the issue until Actions updates to a newer version.
There was a couple things I couldn't fix in that commit, which was UKey2 failing to build because of one of its dependencies failing to build and the same issue with abseil-cpp this project has, and securemessage also failing to build because of its gtest dependency failing to build. Until those get handled in that repository I am unable to create a commit that will fix that here.

I also had a question about development environments that your team is using. I have used a Clang based environment which relies on a compilation database file (compile_commands.json) in order to get syntactical highlighting and similar stuff, although I've come to find out it is difficult to get one of those files using Bazel. My question was what does your team use?

**Comment by proatgram** [comment]
With recent commits to UKey2 that removed submodules and and moves it to Bazel it now compiles perfectly fine! In addition to that, new commits introduced in this repository that brought submodules up-to-date fix the other issues that were occurring with the submodules.

Currently, with just the modifications to those three files I mentioned, building is completely successful on systems with the latest Clang and GCC!

**Comment by proatgram** [comment]
I have updated the branch on my fork to reflect those changes, also including the changes in those files I mentioned, and everything compiles fine. Should I go ahead and create a PR for those couple of include changes here? [This](https://github.com/proatgram/nearby/tree/fix-clang15-gcc13-headers) is the relevant branch that the PR would be based on.

**Comment by bourdakos1** [comment]
Don’t hesitate to create a PR, they are always welcome :)

Regarding developer environment, we mainly use VS Code along with some internal build tools. I think there are a couple Bazel related VS Code plugins that should give syntax highlighting and the likes.

**Comment by proatgram** [comment]
Alright! In that case PR #1683 has been created which when merged will resolve this issue!

Thank you for the info, I will take a look into that! :)

**Comment by bourdakos1** [comment]
Marking as fixed by #1683

Thank you for your contribution 😁

Feel free to re-open this issue if there are still problems related to gcc/clang version

**Comment by proatgram** [comment]
No problem! :smile: 

This issue might be prevalent in other projects in this repository and not just Nearby Connections, but I haven't gotten around to trying to build them. Once I figure out how I will do some more testing.
